File: src/preprocess.py
import homology
import os
import copy

def sort_neighborhoods(adj):
    for u in adj:
        adj[u].sort()

def is_subset(adj, v, w):
    # Check if N(v) ⊆ N(w) using two-pointer
    A, B = adj[v], adj[w]
    i = j = 0
    while i < len(A) and j < len(B):
        if A[i] == B[j]:
            i += 1
            j += 1
        elif A[i] > B[j]:
            j += 1
        else:
            return False
    return i == len(A)

def remove_vertex(adj, v):
    # Remove v from neighbors, then delete its list
    for u in adj[v]:
        # adj[u] is sorted; binary search for v and pop
        lo, hi = 0, len(adj[u])
        while lo < hi:
            mid = (lo+hi)//2
            if adj[u][mid] < v:
                lo = mid + 1
            else:
                hi = mid
        if lo < len(adj[u]) and adj[u][lo] == v:
            adj[u].pop(lo)
    del adj[v]

def preprocess(adj):
    # Deep copy to preserve original input
    adj = copy.deepcopy(adj)

    # Convert adjacency lists to sets for fast subset testing
    neighbors = {v: set(adj[v]) for v in adj}
    removed = set()

    changed = True
    while changed:
        changed = False
        for v in list(neighbors.keys()):
            if v in removed:
                continue
            Nv = neighbors[v]
            for w in neighbors:
                if w == v or w in removed:
                    continue
                Nw = neighbors[w]
                if len(Nw) < len(Nv):
                    continue
                #Allow domination if N(v) ⊆ N(w) ∪ {w} to count self-loops
                if Nv <= (Nw | {w}):
                    # Remove v and its edges
                    for u in Nv:
                        neighbors[u].discard(v)
                    removed.add(v)
                    del neighbors[v]
                    changed = True
                    break
            if changed:
                break

    # Convert sets back to sorted lists for adjacency list format
    reduced_adj = {v: sorted(list(neighbors[v])) for v in neighbors}
    return set(reduced_adj.keys()), reduced_adj

File: src/julia/preprocess.jl
include("homology.jl")

# Little helper function for sorting, just to reduce bulk.
function sort_neighborhoods!(adj)
    for u in keys(adj)
        sort!(adj[u])
    end
end

function is_subset(adj, v, w)
    # Check if N(v) ⊆ N(w) using a two-pointer approach for speed.
    A, B = adj[v], adj[w]
    i = j = 1
    while i <= length(A) && j <= length(B)
        if A[i] == B[j]
            i += 1
            j += 1
        elseif A[i] > B[j]
            j += 1
        else
            return false
        end
    end
    return i > length(A)
end

function remove_vertex!(adj, v)
    # Remove v from neighbors, then update adjacencies
    for u in adj[v]
        # adj[u] is sorted, so use binary search to find and delete v
        idx = searchsortedfirst(adj[u], v)
        if idx <= length(adj[u]) && adj[u][idx] == v
            deleteat!(adj[u], idx)
        end
    end
    delete!(adj, v)
end

function preprocess(adj)
    # Deep copy to preserve original input
    adj_copy = Dict(v => copy(neighbors) for (v, neighbors) in adj)

    # Convert adjacency lists to sets for fast subset testing
    neighbors = Dict(v => Set(adj_copy[v]) for v in keys(adj_copy))
    removed = Set()

    changed = true
    while changed
        changed = false
        vertices_to_check = collect(keys(neighbors))
        for v in vertices_to_check
            if v in removed
                continue
            end
            Nv = neighbors[v]
            for w in keys(neighbors)
                if w == v || w in removed
                    continue
                end
                Nw = neighbors[w]
                if length(Nw) < length(Nv)
                    continue
                end
                # Allow domination if N(v) ⊆ N(w) ∪ {w} to count self-loops
                if Nv ⊆ (Nw ∪ Set([w]))
                    # Remove v and its edges
                    for u in Nv
                        if haskey(neighbors, u)
                            delete!(neighbors[u], v)
                        end
                    end
                    push!(removed, v)
                    delete!(neighbors, v)
                    changed = true
                    break
                end
            end
            if changed
                break
            end
        end
    end

    # Convert sets back to sorted arrays for adjacency list format
    reduced_adj = Dict(v => sort(collect(neighbors[v])) for v in keys(neighbors))
    return Set(keys(reduced_adj)), reduced_adj
end